### ISSUE TITLE
Add Monitoring Plugin

### ISSUE THREAD

**Issue by sradco** [kubernetes-sigs/kubebuilder#3078]
### What do you want to happen?

We would like to add a new monitoring plugin that will help operators developer with setting up Prometheus based monitoring for their operator, will provide them with best practices and tooling to shorten their time to get up to speed with monitoring requirements and help with standardizing the way monitoring is implemented in operators.

The proposed structure and content:

```
operator main/
  monitoring/
    READM.md
    metrics/
      metrics.go (script to register the metrics and additional info)
      example_metrics.go  (script for creating the custom metrics and additional info)
      util/
        util.go  (Help functions and other utils for handling the metrics)
    alerts/
      alerts.go  (script for creating alerts and additional info)
      util/
        util.go  (Help functions and other utils for handling metrics)
    dashboards/ (Here the Grafana dashboard for controller-runtime metrics can be added if the users will add an additional command)
    tools/
      metricsdocs.go
    prom-rule-ci/
      prom-rules-tests.yaml (script for unit tests based on Prometheus tools)
  docs/
    monitoring/
      metrics.md
      runbooks/
       Runbook.md (Runbook template for helping with understanding and mitigating alerts)
```

PR https://github.com/operator-framework/operator-sdk/pull/5975 is also related and can be referenced to provide additional best practices. It is correctly on hold since we would like to replace the examples with the examples here.

**Comment by Kavinjsir** [comment]
The idea sounds great 👍🏼 .
Looking forward to have this feature.
(I'd be happy to help with if possible 😃 )

**Comment by camilamacedo86** [comment]
HI @sradco, 

Interesting. Could you please let us know more about your proposal? 
What would this plugin would scaffold? Have we an example of what would you like to scaffold within?

**Comment by umangachapagain** [comment]
@sradco  IMO `a monitoring plugin that will help operators developer with setting up Prometheus based monitoring` sounds much bigger in scope than the implementation.

When I first read the issue, I was expecting a scaffold that will help deploy Prometheus stack (creating Prometheus resources). But, this is more about generating metrics (which may or may not be consumed by Prometheus). So, we should probably consider renaming the issue and plugin name to be more specific to metrics.

And since we're talking about best practices, will there be a recommended set of metrics that all operators should implement?

I like the overall idea.

**Comment by sradco** [comment]
@camilamacedo86 @umangachapagain  I updated the description of the issue. The plugin is meant to help the operators developers with the on boarding process of adding monitoring(metrics, alerts, and more) and to offer the best practices to do that correctly. 

For example, having the metrics code logic in  /monitoring/metrics and not inside the core operator code. In the core operator code there should only be a call to update the metric.

**Comment by Kavinjsir** [comment]
I have the same sense as @umangachapagain [pointed out](https://github.com/kubernetes-sigs/kubebuilder/issues/3078#issuecomment-1330042213). As an operator author, I would assume to have a whole monitoring stacks solution for the first glance on the name `monitoring plugins`.

When going over for details, it feels more like `prometheus plugin` where it mainly provides approaches to define: metrics, recording rules, alert rules that are all "prometheus scoped".

Talking about monitoring, I personally am thinking of metrics, logging, tracing data, that is beyond what this issue want to bring about.

However, the idea overall looks good. Since populating customized metrics can be a common case, having a general approach to simplify metrics/rules implementation do save developers' time.

In my own opinion, I would expect this approach to be flexibly imported in my project. Which means, my project might not necessarily to be a scaffold by Kubebulider. From there, would it be possible if we change it to be a library or a separate tools like `controller-gen`?

**Comment by varshaprasad96** [comment]
@sradco We had a discussion regarding the monitoring plugin during the community meeting and a few insights were brought up by @camilamacedo86 and @Kavinjsir. I'm summarizing the list of concerns below:
1. By @camilamacedo86 @Kavinjsir - Since the functions and methods are generic, does it make sense to have this as a library rather than a plugin?
2. By @Kavinjsir - With "monitoring" seems misleading, since monitoring just does not involve prometheus. It includes tracing, logging and other methods. Whereas this plugin just talks about only prometheus.
3. By @Kavinjsir - This seems to be a golang specific plugin. The best practices mentioned (for adding alters, declaring a custom metric etc) here are relevant only for go based operators, and not for a java based operator project. Considering this,  does this fit more in a prometheus project than in kubebuilder?

I would like to add like to add some of my thoughts inline, but I also defer to @sradco and team for their point of view.

> 1. By @camilamacedo86 @Kavinjsir - Since the functions and methods are generic, does it make sense to have this as a library rather than a plugin?

**Comment by Kavinjsir** [comment]
@varshaprasad96 Big thanks for bringing what we've discussed in weekly with details! 👍🏼 

@sradco  I would also like to add my thoughts to supplement the points above:

For **Q1**
> What in the plugin would be helpful to users is: https://github.com/kubernetes-sigs/kubebuilder/blob/28fe31ee240090c2009e5c0a6611d3f0b85407f1/testdata/project-v3-declarative-v1/monitoring/tools/metricsdocs.go - wherein I mention the custom metrics and have a standard documentation generated for it. Whether this warrants a separate plugin or just an additional script in a discussion.

I'm also happy to see a documentation with introduction of custom metrics if existing.
Just like how we have [docs of default metrics](https://book.kubebuilder.io/reference/metrics-reference.html#default-exported-metrics-references) that helps us to identify the metrics for our needs.

For **Q2**
Since what this issue and the corresponding PR are discussing is closely related to Prometheus, maybe we can consider if a better naming can be possible?

For **Q3**
> ...because eventually when this is moved to an external plugin it can just be considered as an extension for having best practices for go operator.

I'm good if such feature becomes Golang specific finally.
Well, one point I'm personally thinking is: except `metrics`, other prometheus manifests may be not necessarily go-specific.
Like the implementation and usage of `alerting rules` and `recording rules`.

---

In fact, I would imagine that the default metrics provided by controller-runtime may cover many use cases when monitoring on API reconciliation.
So wondering if it is possible that "implementing custom metrics for CRD" may become a comparatively advanced topic?
And before that, the users might be more interested in utilizing the existing metrics to create "recording rules" / "alerting rules"?

Actually, many projects provides mixins to give these prometheus resources "off the shelf", for instance:
- [kubernetes-mixins](https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/master/alerts/resource_alerts.libsonnet)
- [kube-prometheus](https://github.com/prometheus-operator/kube-prometheus/blob/v0.11.0/manifests/prometheus-prometheusRule.yaml)
- [jaeger-mixin](https://github.com/jaegertracing/jaeger/blob/main/monitoring/jaeger-mixin/alerts.libsonnet)

So as an operator user, I might also be happy if the plugin/lib can bring me similar manifests. (Just one thought around monitoring...)

**Comment by camilamacedo86** [comment]
Hi @sradco,

After think more about that, I'd like to share the following suggestions:

The utils proposed in the plugin does not seems to fit under a good practices to be scaffolded in the projects. I think that would to be a lib. Since it is very prometheus specific and not "K8s" context I think that it might be accepted under `github.com/prometheus/` 

Then, If we remove the utils from the plugin we have few code lines and the readme. So, that does not seems bring value for one plugin and Kubebubuilder users. _PS.: But is not a README a good idea and we scaffold in the go/v*? Yes, it is nice but we do **not** have the go/v* plugins to scaffold a README. it is just a nit and that is not what address their value._ 

**Suggestion**

> By latest, see that operator-sdk has a best practice section (https://sdk.operatorframework.io/docs/best-practices/) and scaffold samples within code which is used in the tutorials. Also, has samples with utils and related "good practices" examples addressed as part of your prevous initiatives (https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/monitoring/memcached-operator). So, why not add this code there?

**Conclusion**

IMO we should not accepted it in Kubebuilder _(it does seem address value for the project and seems to be too prometheus specific which is out of the scope)._ 

The goal of Kubebuilder is not accepted and/or have a lot of plugins because it would be hard to keep maintained. Therefore, Kubebuilder has been investing efforts in the [plugin phase 2](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/extensible-cli-and-scaffolding-plugins-phase-2.md) to provide an API to allow the community created its [own plugins](https://book.kubebuilder.io/plugins/creating-plugins.html) that would not fit in its space and use as external plugins (see https://github.com/kubernetes-sigs/kubebuilder/issues/2600). We really would like to provide only few plugins shipped with the CLI and invest efforts in the API. 

So, IMO I do not think that we can accept this proposal and my vote is -1 (here).Also, I understand that the others thoughts go along with. So, are we you ok with we close this issue and the open pr as defered? 

c/c @Kavinjsir @varshaprasad96 @umangachapagain

**Comment by umangachapagain** [comment]
> 
> > The goal of Kubebuilder is not accepted and/or have a lot of plugins because it would be hard to kept maintained. Therefore, Kubebuilder has been investing efforts in the [plugin phase 2](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/extensible-cli-and-scaffolding-plugins-phase-2.md) to provide an API to allow the community created its [own plugins](https://book.kubebuilder.io/plugins/creating-plugins.html) that would not fit in its space and use as external plugins (see #2600). We really would like to provide only few plugins shipped with the CLI and invest efforts in the API.
> 
> Also, I understand that the others thoughts go along with. **So, are we you ok with we close this issue and the open pr as not accepted?**
> 
> @Kavinjsir @varshaprasad96 @umangachapagain
>
I'd still like to see this as a plugin. External plugin system seems to be the right fit. Is there a centralized repo for community contributed plugins?

@camilamacedo86  Operator SDK already scaffolds some resources specific to Prometheus. Do you know if SDK does it on it's own or gets it via KB? As you suggested, SDK seems like a right fit to integrate such a plugin.

I envision this as an external KB plugin which is leveraged by SDK as a default. SDK already brings in ServiceMonitor. This plugin can add metrics and PrometheusRule to it and document best practices.

**Comment by camilamacedo86** [comment]
Hi @umangachapagain, 

> I'd still like to see this as a plugin. External plugin system seems to be the right fit. Is there a centralized repo for community contributed plugins?

Why? What advantages a plugin that mainly only scaffolds a README can bring to you vs the cost to keep it maintained?
Do you need a plugin for that? Should not a doc and a sample/example enough for you achieve the goal?

> @camilamacedo86 Operator SDK already scaffolds some resources specific to Prometheus. Do you know if SDK does it on it's own or gets it via KB? As you suggested, SDK seems like a right fit to integrate such a plugin.

The only specific prometheus thing scaffold in the projects are the rule to allow enable the default metrics because by default they are protected. 

**No **utils** code in any form are scaffolded in the projects done by SDK or Kubebuilder because that fits to a lib or into a third-party tool that can be consumed by the projects usually.**

My suggestion was check if this code cannot be added into the sample/example https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/monitoring/memcached-operator (not default scaffolds or plugin) since there has other "good practices" regards this topic and it can be used into the SDK docs.

**Comment by sradco** [comment]
> Hi @umangachapagain,
> 
> > I'd still like to see this as a plugin. External plugin system seems to be the right fit. Is there a centralized repo for community contributed plugins?
> 
> **Why? What advantages a plugin that mainly only scaffolds a README can bring to you vs the cost to keep it maintained? Do you need a plugin for that? Should not a doc and a sample/example enough for you achieve the goal?**

I disagree. As I mentioned before, this plugin is meant for an easier onboarding to Prometheus monitoring. This PR is the initial work that includes the metrics part. It doesn't only include README but includes the example files that the developers will add their specific metrics to.

This is a very important part that will greatly help operator developers, with adding new metrics.
Also this integrates an important best practice which is to separate the monitoring code from the core operator code. 
This is emphesised both in the example files and the README. It adds tooling that help them to document the operator specific metrics automatically. This is only the beginning.

For alerts we will also add a second PR that adds the required files for adding the alert rules, recording rules, prometheus unit tests, runbooks.

We plan to complement this plugin with a best practices doc that about metrics naming conventions, alerts labels, which are correctly not documented in a central place and its the wild west which makes it hard for end users to make good use of the monitoring stack.

I agree that this maybe specific to Prometheus stack, but the Grafana plugin is specific to Grafana and its in Kubebuilder. I don't mind making this a Prometheus stack plugin.

Prometheus is a highly used stack and it deserves the attention. The Grafana plugin can become part of it.


> 
> > @camilamacedo86 Operator SDK already scaffolds some resources specific to Prometheus. Do you know if SDK does it on it's own or gets it via KB? As you suggested, SDK seems like a right fit to integrate such a plugin.
> 
> The only specific prometheus thing scaffold in the projects are the rule to allow enable the default metrics because by default they are protected.

What we added to this operator will save a lot of development time, frustration and prevent pitfalls that will later cost in more time to fix.

> 
> **No **utils** code in any form are scaffolded in the projects done by SDK or Kubebuilder because that fits to a lib or into a third-party tool that can be consumed by the projects usually.**

We are not providing a lib. The files are the base for the operator developers to build on.

> 
> My suggestion was check if this code cannot be added into the sample/example https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/monitoring/memcached-operator (not default scaffolds or plugin) since there has other "good practices" regards this topic and it can be used into the SDK docs.

The example dashboard will not have the same affect. We want to give the developers that starting point that will lead them for easier and better implementation of the Prometheus stack monitoring.

The memcached operator will be build on top of this plugin and add the actual metrics, alerts and so on. We still want to add that.

**Comment by camilamacedo86** [comment]
HI @sradco, 

Thank you for your input. Could you attend the next community meeting?
Would be great if we be able to discuss this topic altogether and by voice. WDYT? 

However, please check my comments inline I hope that clarifies the thought and the differences between the grafana plugin and the proposed idea so far. 

However, let's first ensure that we are all on the same page here and that the utils/code generated by the proposed plugin should not be scaffolded in the project itself. Then, after moving this implementation to a lib that will be scaffolded by the plugin? (That is what we need to discuss here, right?)

> It doesn't only include README but includes the example files that the developers will add their specific metrics to.

**Codes examples in Kubebuilder are only addressed via docs/tutorials. 
All plugins scaffold useful/valid code NOT examples.** 

Therefore, following good practices and conventions shows a pre-requirement for any code done by the tool.

_In SDK, that can be found in the test data samples where for example, you have been working on adding content regards this initiative. ([here]( https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/monitoring/memcached-operator)) **so, if the main goal of the plugin is scaffold examples then we cannot accept this one.**_ 
 
> We plan to complement this plugin with a best practices doc that about metrics naming conventions, alerts labels, which are correctly not documented in a central place and its the wild west which makes it hard for end users to make good use of the monitoring stack.

Nomenclature and connections show that usually fit very well in docs and examples but not in tools to generate code. Tools usually generate a valid/useful code following these concepts. 

> I agree that this maybe specific to Prometheus stack, but the Grafana plugin is specific to Grafana and its in Kubebuilder. I don't mind making this a Prometheus stack plugin.

The grafana plugin does not scaffold utils to work with grafana. (such as the proposed idea so far) 
It scaffolds the dashboards to check the default manager exports metrics within. So, we have a valid/useful code generated based on what Controller-Runtime and Kubebuilder do by default. 

So, would possible to change the plugin/proposed idea to address monitoring for what is scaffolded by default? What could be addressed in the default scaffold regards this matter? Could we have alerts for the default metrics? Have any metric scaffold by default in controller-runtime that does not follow this convention and we could improve the implementation?

That is something that IMO could be very nice and would have +1 vote here. _(I mean, without considering any utils code and let's assume that in another place)_

**Comment by sradco** [comment]
I will attend the meeting. 

- Perhaps the word `example` wasn't the exact word for what we mean and what is implemented.

    **The files do not include metrics/alerts examples. They include the actual code** needed for creating and exposing operator metrics and alerts. 
    
    The developers only need to add their custom metrics/alerts/recording rules and so on to them.

- In Memcached operator we use the code and add the Memcached specific metrics/alerts and all other the examples.

- We can rename the "util" folder and file, if it is confusing. Please see what @machadovilaca wrote.

- The Grafana plugin does include example file for adding custom metrics https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-metrics/grafana/custom-metrics/config.yaml.
And its used here https://github.com/kubernetes-sigs/kubebuilder/blob/8069fa5e3a64a1671c7b80ae5dc588567c9aa295/pkg/plugins/optional/grafana/v1alpha/scaffolds/internal/templates/custom.go to generate the template.

    **We believe this is the proper way to add custom metrics.**

**This plugin doesn't aim to add specific metrics and alerts and other specific resources.**
**Adding specific metrics, alerts, dashboards should use this plugin and be added them on top of it, while keeping the best practices**

**Comment by camilamacedo86** [comment]
Hi @sradco,

Following my comments inline. 

 > The files do not include metrics/alerts examples. They include the actual code needed for creating and exposing operator metrics and alerts.

In the Proposed PR, the code generated is lib/bin features. not CLI ones and should not be inside of the "Operator" itself. 
The projects built by Kubebuilder could use them but is not a good practice/arch design to ship them inside of the projects.

> It is the same like https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-multigroup/apis/foo/v1/bar_types.go

That is the result of the code to generate and create APIs with the project. It is not the same, see: 
- Add comments in the code to let users know how to use a feature is acceptable
- Creating a feature that provides only comments and examples is not.

> In Memcached operator we use the code and add the Memcached specific metrics/alerts and all other the examples.
It is a sample provided in the SDK repo not Kubebuilder

It is a sample in SDK. Kubebuilder has no Memcached operator or plugin that generates a specific Memcached operator.

_The sample is generated automatically via a stack available in SDK for maintainability/development purposes and is not part of any CLI feature. Therefore, ihmo what you are trying to add here could fit well in there in the sample but not in a plugin/cli feature._ 

> We can rename the "util" folder and file, if it is confusing. Please see what @machadovilaca wrote.

The problem is not the name but what it does.
The code inside the utils should be centralized in lib and consumed by the projects. The binary also should not be built for all projects produced by the tool. that should be provided outside and consumed by the projects. 

**(example)** 

We do not have the Controller-Runtime features scaffolds in the projects via Kubebuilder.
The projects import controller-runtime, and all code is centralized in the lib. 

We do not scaffold the code to generate the envtest binary in all projects done with Kubebuilder. 
We provide the env test bin, and Kubebuilder's projects use it. 

> The Grafana plugin does include example file for adding custom metrics https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4-with-metrics/grafana/custom-metrics/config.yaml.
And its used here https://github.com/kubernetes-sigs/kubebuilder/blob/8069fa5e3a64a1671c7b80ae5dc588567c9aa295/pkg/plugins/optional/grafana/v1alpha/scaffolds/internal/templates/custom.go to generate the template.

Grafana plugin provides a feature where you can input the info about your custom metrics, generating a valuable, final useful result that is Grafana Dashboards 100% done to be deployed on the cluster to check its metrics.

Then, again is great to have commands with examples in the code to improve the usability of a feature
However, I do not agree with creating features to provide examples. Examples can be addressed via tutorials and etc.

> We believe this is NOT the proper way to add custom metrics.

Grafana plugins do NOT add custom metrics The Grafana plugin allows you to say, "here are my custom metrics" please generate "Garafana dashboards for them". That is a valid feature/CLI option. 

> This plugin doesn't aim to add specific metrics and alerts and other specific resources.
Adding specific metrics, alerts, dashboards should use this plugin and add them on top of it, while keeping the best practices

Sorry, but I could not follow this one. However, if you disagree with the grafana plugin, then that should be addressed
by creating an issue to discuss what it does today and what you would like that the plugin was capable of todo or change on it and not propose another plugin instead. 

I hope that makes sense and can properly clarify the points.

**Comment by varshaprasad96** [comment]
@camilamacedo86 I understand your point of view, and to summarize adding two important points here that are creating questions on whether this issue should be accepted or not. I'll add my answers to them inline:

Question 1: Are the helpers scaffolded by the plugin valuable? What are the benefits of scaffolding this inside an operator project, can these be just helpers in a library. 
Question 2: Do we accept this as a plugin in KB? If we accept this as an plugin in KB, how do we manage the engineering effort to maintain it? 

Answers inline:

Question 1:

By definition, plugins are extensions to KB cli, which address "specific" needs of the user. These needs can either be addressed by scaffolding or generating code that is helpful to the user, like the grafana plugin does, or by enforcing a best practice (like the deploy image plugin does). Either ways until a plugin "addresses the needs of the user" it is valuable. The monitoring plugin here, is trying to enforce best practices by directing the user on where or how to add metrics in their project following best practice. After digging up, I understand that "util.go" is not a library, but is a helper which an operator author should eventually have "inside their project". Let's take an example here:

The method `MustRegister` (https://github.com/prometheus/client_golang/blob/254e5468413f19fb75cdad45f5ddc0b8c975188c/prometheus/registry.go#L177) in very simple terms registers a collection of metrics (in promethteus' terms "collector"). This is an API provided by the user to register their custom metrics. Hence prometheus provides this as a library. An operator author has to use this method in their respective project to register their custom metrics. What  usually folks do is add this in their operator code (usually in reconciler) which does work without any problems but is unnecessary and "NOT" a good practice. What the monitoring plugin is trying to do is scaffold this piece of code in a separate folder and enforce the users to organize the monitoring pieces of their project in the right place. Having the wrappers like this - https://github.com/kubernetes-sigs/kubebuilder/pull/3106/files#diff-75224dd6164f9903e8acf72b976de2f46f97c451a3b784fcd1d20347a90ffcbaR29 in a library is adding no value, the user will still continue to make the same mistake of having them at the wrong location. 

This is similar to the deploy-image plugin - without this everything works as expected, but the operator project does not follow a best practice. A user can have an unfiltered cache, their project would still work, but having cache selectors will enable the operator to be more efficient - which is one of the things that deploy-image does.

Imo this should be considered as a plugin. Having the helpers in a library would add no value, its like creating wrapper around prometheus APIs which is of no need. The motive of this plugin is to enforce a best practice and having these in a library would defeat that purpose, since users would not even use the library which is created from this.

Question 2: Do we accept this is a plugin in KB?

I agree with @camilamacedo86's point of view in this aspect wherein the eventual effort of maintaining the plugin inside KB can be painful. Having a plugin as a part of KB has a long list of problems (which I wouldn't repeat here), but this is why Phase 2 was introduced. Having out-of-tree plugins solves question 2, where as KB community we need not be worried of maintaining the code base. It would be external, can be used with KB, but will have its own set of maintainers and release cycle. This issue was brought up in community meeting, however, the idea was to start this as a phase 1.5 and eventually move it to phase 2 plugin, since as KB maintainers we haven't done a good job of creating a step-by-step documentation on building phase 2 plugins. However, if @sradco and team are comfortable on making this phase 2, that would definitely be the best solution.

Tl;dr; the first step is to answer whether we can consider this to be a plugin or not. (Imo we should for which I have provided a detailed explanation above). The next is if this should be phase 1.5 or a phase 2 plugin. If the authors are comfortable with making this external using phase 2, then definitely +1 for it! 

cc: @camilamacedo86 @sradco @machadovilaca @umangachapagain @Kavinjsir

**Comment by Kavinjsir** [comment]
I'm would also vote for `external plugin`:
1. The proposed "best practices" looks great especially in case the author continuously defines number of metrics. If there are just a few metrics, will this practice be heavy?
2. Instead of "guide users to generate metrics for operator", does the code scaffolded under `monitoring/` look more like "guide users to generate metrics in go project"?

I'm wondering if there is some specific `metric practices` that is operator scoped, because Kubebuilder, as I suppose, is the tool to help user create a good operator project (versus "golang project").

**Comment by camilamacedo86** [comment]
HI @varshaprasad96,

Thank you for all your input. Following some comments inline

**Regards the question: Question 1: Are the helpers scaffolded by the plugin valuable? What are the benefits of scaffolding this inside an operator project, can these be just helpers in a library.**

See that the utils also have a binary to build the docs for the metrics. In the same way and for reasons that we do not scaffold a code to build envtest bin inside of the project, I do not agree that is something that should be scaffolded by default. 

Therefore, I was also wondering why it is required to implement and build a binary. If that is a good practice, does it not exist for projects that already provide the docs for the metrics? Are not the metrics implement in go so godoc will not generate the docs for it? 

**Regards**
> This is similar to the deploy-image plugin - without this everything works as expected, but the operator project does not follow a best practice.

I think we cannot compare this proposal with the deploy image plugin at all because the deploy image delivers a result which is its main goal. "I want to create a project to deploy/manage an image so that I can use the plugin to scaffold the code for me that can be used, and then if I want, I can customize and do my changes on top of that." 

Following the goods practices to delivery, the result is a pre-requirement for any plugin/Kubebuilder scaffolds, not any end goal or primary purpose. Yes, one of the nice things about deployImage plugin is showing how to implement the k8s conventions and etc but the primary purpose is to ensure the scaffold for a common scenario. See its mainly motivation form where the idea began https://github.com/operator-framework/operator-sdk/pull/2158 (the goal was to provide the scaffold to speed up)


**Conclusion**

**I think we can all convey that we reached a consensus that this proposal does fit well for KB and cannot be accepted.** 

_(ihmo) I do not think nomenclature, examples, or conventions are not good, mainly end goals for plugins and would be addressed better via docs and examples. Plugins are features and should address a need/requirement. Ensure that the code resulted from the CLI/feature follows good practices, etc is a pre-requirement and does not fit as "that is the reason for this plugin exist". However, the idea with plugins phase 2 is for Kubebuilder to provide the libs (which are implemented already. It is only missing docs and e2e test to ensure the implementation), and the community can use them as the API as please them._ 

@sradco, if you still not agreeing with, we close this one as deferred and want to first have a chance to discuss it in the next community meeting by voice; we can leave it open until there. Otherwise, please let us know if you are ok with us closing this one now. 

@Kavinjsir @umangachapagain @everettraven

**Comment by varshaprasad96** [comment]
@camilamacedo86 Just adding my follow up thoughts here.

> Therefore, I was also wondering why it is required to implement and build a binary. If that is a good practice, does it not exist for projects that already provide the docs for the metrics? Are not the metrics implement in go so godoc will not generate the docs for it?

The docs are not the only files scaffolded. The plugin brings in the code required to setup metrics as well (https://github.com/kubernetes-sigs/kubebuilder/pull/3106/files#diff-75224dd6164f9903e8acf72b976de2f46f97c451a3b784fcd1d20347a90ffcbaR29). If an operator author wants to add metrics using prometheus, then its upto them to use this plugin and scaffold the helpers and its documentation (its not by default). Envtest imo is a different case, where we use the binary (as is) to spin up the test cluster. The users of the operator author project need not know what binaries are downloaded by envtest, or what is the use of "setup-envtest", or even the use of each binary. But the users of these metrics need to know how it is registered as well as what each one of them do. 

In my mind, the documentation can be of two kinds:
1. Docs which help operator author to generate metrics - This should definitely be in a common place as KB docs (like in KB book). =>  (i.e its like KB is the product, and customer are operator authors)
2. Docs which help the users of operator - If these are operator specific docs then having it in the project does add value to its users => (i.e operator author is delivering the operator image as the product, and users are the customers).

I agree that docs need not be a part of the operator image (or the binary built for it), but can definitely be a part of the project, for which this plugin is helpful. Just the helpers used to register metrics can be in `pkg/`.

> I think we can all convey that we reached a consensus that this proposal does fit well for KB purposes and views and, unfortunately, cannot be accepted to be shipped within.

(please correct me if I'm wrong) Do we agree that this could be a plugin? just that it cannot be a phase 1.5 plugin and shipped along with KB. 

I'm more concerned on opinions on why this shouldn't be a plugin than whether it should be in KB (as phase 1.5) or an external one. If the latter is true (this can be plugin but external), then we can surely close this issue and encourage the authors to create issues specifically with respect to phase 2 as they start building on it.

**Comment by camilamacedo86** [comment]
Hi @varshaprasad96, 

IHMO:

**Regards the specific propose of this plugin**

After we remove the [testdata/project-v3-declarative-v1/monitoring/tools/metricsdocs.go](https://github.com/kubernetes-sigs/kubebuilder/pull/3106/files#diff-455f7928933e227aa87e92a8cc46759fc8fd4ea3ca27cbd08bb80af22a858156) which we all agree that is not a good practice be scaffolded in the project (Not sure why the binary is indeed required, are not godocs doing that already? Have not projects that can do it? Could not it be addressed in another project ? )

The goal of the plugin still:

- scaffold a README to doc the metrics 
- scaffold a wrapper for Prometheus client and provide an example over how/where according to them would the metrics be better organized

Then, let's think about the value:

- Custom metrics is not a common need. I cannot think in a good example where it would be required. (seems to me address very specific use cases (let's guess here < 20% ))
- By using the plugin it will only try to enforce a "convention/standard" that they would like to promote and not delivery something that I can check working and get an immediately result. 
- It seems that could be double if that is or not the best away to achieve the goal 
- Nomenclature, conventions are usually better described by docs and examples. 

So, in POV: 

The work required to maintain a plugin does not pay off what it can be return and the same goal can be achieved by examples and docs which seems fit better.

**Now, about what plugins Kubebuilder could accept or not to be shipped with and maintained by the project?*

I think we need to keep in mind only accept plugins that can bring value for the common scenarios and what is done by default (i.e grafana plugin scaffolds the dashboards for what is done by default, deployImage address a very common scenario). 

Looking for maintainability, I would like to indeed suggest we review for example the declarative plugin because if we have no adoption from the community on it then we could remove this one as well and let it be addressed as an external plugin which could leave in the declarative repo. Therefore, be kept maintained for those that have been maintained this project are more skilled about, will be able to better improve the plugin as address what is new one that.

I love the idea of the plugin but if they does not address common case scenarios, does not provide value over what is done by default and only can be helpful for edge cases then, I do not think that we can have then in Kubebuilder because that would result in something like "HelmChartsRepo" with so many of them that would be impossible to manage.

**Comment by camilamacedo86** [comment]
Hi @varshaprasad96 @sradco @Kavinjsir @everettraven,

It seems that the recommendation will be addressed via examples and docs in operator-sdk side. If so, do we agree in close this one as the PR:  https://github.com/kubernetes-sigs/kubebuilder/pull/3106?

**Comment by everettraven** [comment]
@camilamacedo86 I think it is fine to close #3106 as per [this comment](https://github.com/kubernetes-sigs/kubebuilder/pull/3106#issuecomment-1357621972)

I think closing this issue is fine unless we feel there needs to be more discussion about making it an external plugin.

**Comment by sradco** [comment]
@camilamacedo86 I think we can close this for now.
I still believe that a plugin for monitoring would be very valuable, but since it will not be accepted as part of the kubebuilder code, we are considering the amount of effort and time this would require from our team.

Having the documentation and examples is a good starting point. It can also be expended to include additional information about logs, tracing, events etc and we plan to help in adding additional information as we go.

From my POV, adding the plugin would allow automations that would save development time and on boarding time, will allow code reusability, improve the code quality and would ultimately result in better user experience.
